File: app/src/main/java/com/ibashkimi/lockscheduler/adapter/ProfileAdapter.java
package com.ibashkimi.lockscheduler.adapter;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.ibashkimi.lockscheduler.R;
import com.ibashkimi.lockscheduler.addeditprofile.conditions.wifi.SelectableAdapter;
import com.ibashkimi.lockscheduler.databinding.ItemProfileBinding;
import com.ibashkimi.lockscheduler.model.Profile;
import com.ibashkimi.lockscheduler.model.action.LockAction;
import com.ibashkimi.lockscheduler.model.condition.PlaceCondition;
import com.ibashkimi.lockscheduler.model.condition.PowerCondition;
import com.ibashkimi.lockscheduler.model.condition.TimeCondition;
import com.ibashkimi.lockscheduler.model.condition.WifiCondition;
import com.ibashkimi.lockscheduler.util.ConditionUtils;

import java.util.List;


public class ProfileAdapter extends SelectableAdapter<ProfileAdapter.ProfileViewHolder> {

    private List<Profile> mProfiles;
    private Callback mItemListener;

    public ProfileAdapter(List<Profile> profiles, @NonNull Callback listener) {
        this.mProfiles = profiles;
        this.mItemListener = listener;
    }

    @NonNull
    @Override
    public ProfileViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        ItemProfileBinding binding = ItemProfileBinding
                .inflate(LayoutInflater.from(parent.getContext()), parent, false);
        return new ProfileViewHolder(binding, getItemListener());
    }

    @Override
    public void onBindViewHolder(@NonNull final ProfileViewHolder holder, int position) {
        final Profile profile = getProfiles().get(position);
        holder.init(profile);
        holder.setSelected(isSelected(position));
    }

    @Override
    public int getItemCount() {
        return mProfiles.size();
    }

    public void setData(List<Profile> data) {
        this.mProfiles = data;
        notifyDataSetChanged();
    }

    public List<Profile> getProfiles() {
        return mProfiles;
    }

    private Callback getItemListener() {
        return mItemListener;
    }

    public interface Callback {

        void onProfileClick(int position);

        void onProfileLongClick(int position);
    }

    static class ProfileViewHolder extends RecyclerView.ViewHolder {
        ItemProfileBinding binding;

        Callback listener;

        ProfileViewHolder(ItemProfileBinding binding, @NonNull Callback listener) {
            super(binding.getRoot());
            this.binding = binding;
            binding.getRoot().setOnClickListener(view -> listener.onProfileClick(getAdapterPosition()));
            binding.getRoot().setOnLongClickListener(view -> {
                listener.onProfileLongClick(getAdapterPosition());
                return true;
            });

            this.listener = listener;
        }

        void init(@NonNull Profile profile) {
            if (!profile.getName().equals("")) {
                binding.name.setText(profile.getName());
            } else {
                binding.name.setVisibility(View.GONE);
            }
            LockAction enterLockAction = profile.getEnterExitActions().getEnterActions().getLockAction();
            if (enterLockAction == null)
                throw new IllegalArgumentException("Profile " + profile.getName() + " doesn't contain an enter lock action.");
            binding.enterLockMode.setText(enterLockAction.getLockMode().getLockType().getValue());
            LockAction exitLockAction = profile.getEnterExitActions().getExitActions().getLockAction();
            if (exitLockAction == null)
                throw new IllegalArgumentException("Profile " + profile.getName() + " doesn't contain an exit lock action.");
            binding.exitLockMode.setText(exitLockAction.getLockMode().getLockType().getValue());
            PlaceCondition placeCondition = profile.getConditions().getPlaceCondition();
            if (placeCondition != null) {
                binding.placeSummary.setText(placeCondition.getAddress());
                binding.locationLayout.setVisibility(View.VISIBLE);
            } else {
                binding.locationLayout.setVisibility(View.GONE);
            }
            TimeCondition timeCondition = profile.getConditions().getTimeCondition();
            if (timeCondition != null) {
                binding.days.setText(ConditionUtils.daysToString(itemView.getContext(), timeCondition));
                binding.interval.setText(ConditionUtils.internvalToString(timeCondition.getStartTime(), timeCondition.getEndTime()));
                binding.timeLayout.setVisibility(View.VISIBLE);
            } else {
                binding.timeLayout.setVisibility(View.GONE);
            }
            WifiCondition wifiCondition = profile.getConditions().getWifiCondition();
            if (wifiCondition != null) {
                CharSequence[] wifiList = new CharSequence[wifiCondition.getWifiList().size()];
                for (int i = 0; i < wifiList.length; i++)
                    wifiList[i] = wifiCondition.getWifiList().get(i).getSsid();
                binding.wifiSummary.setText(ConditionUtils.concatenate(wifiList, ", "));
                binding.wifiLayout.setVisibility(View.VISIBLE);
            } else {
                binding.wifiLayout.setVisibility(View.GONE);
            }
            PowerCondition powerCondition = profile.getConditions().getPowerCondition();
            if (powerCondition != null) {
                binding.powerLayout.setVisibility(View.VISIBLE);
                binding.powerSummary.setText(powerCondition.getPowerConnected() ? R.string.power_connected : R.string.power_disconnected);
            }
        }

        void setSelected(boolean selected) {
            binding.cover.setVisibility(selected ? View.VISIBLE : View.GONE);
        }
    }
}

File: app/src/main/java/com/ibashkimi/lockscheduler/adapter/ProfileAdapter.kt
package com.ibashkimi.lockscheduler.adapter

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.ibashkimi.lockscheduler.R
import com.ibashkimi.lockscheduler.adapter.ProfileAdapter.ProfileViewHolder
import com.ibashkimi.lockscheduler.addeditprofile.conditions.wifi.SelectableAdapter
import com.ibashkimi.lockscheduler.databinding.ItemProfileBinding
import com.ibashkimi.lockscheduler.extention.concatenate
import com.ibashkimi.lockscheduler.model.Profile
import com.ibashkimi.lockscheduler.model.action.LockAction
import com.ibashkimi.lockscheduler.model.condition.*
import com.ibashkimi.lockscheduler.model.find
import com.ibashkimi.lockscheduler.model.findCondition
import com.ibashkimi.lockscheduler.util.intervalToString
import com.ibashkimi.lockscheduler.util.toDaysString
import java.util.*

class ProfileAdapter(var profiles: MutableList<Profile>, private val itemListener: Callback) :
    SelectableAdapter<ProfileViewHolder?>() {
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ProfileViewHolder {
        val binding = ItemProfileBinding
            .inflate(LayoutInflater.from(parent.context), parent, false)
        return ProfileViewHolder(binding, itemListener)
    }

    override fun onBindViewHolder(holder: ProfileViewHolder, position: Int) {
        val profile = profiles[position]
        holder.init(profile)
        holder.setSelected(isSelected(position))
    }

    override fun getItemCount(): Int {
        return profiles.size
    }

    fun setData(data: MutableList<Profile>) {
        profiles = data
        notifyDataSetChanged()
    }

    interface Callback {
        fun onProfileClick(position: Int)
        fun onProfileLongClick(position: Int)
    }

    class ProfileViewHolder(
        var binding: ItemProfileBinding,
        listener: Callback
    ) : RecyclerView.ViewHolder(binding.root) {
        var listener: Callback
        fun init(profile: Profile) {
            if (profile.name != "") {
                binding.name.text = profile.name
            } else {
                binding.name.visibility = View.GONE
            }
            val (lockMode) = profile.enterActions.find<LockAction>()
                ?: throw IllegalArgumentException("Profile " + profile.name + " doesn't contain an enter lock action.")
            binding.enterLockMode.text = lockMode.lockType.value
            val (lockMode1) = profile.exitActions.find<LockAction>()
                ?: throw IllegalArgumentException("Profile " + profile.name + " doesn't contain an exit lock action.")
            binding.exitLockMode.text = lockMode1.lockType.value
            val placeCondition: PlaceCondition? = profile.findCondition()
            if (placeCondition != null) {
                binding.placeSummary.text = placeCondition.address
                binding.locationLayout.visibility = View.VISIBLE
            } else {
                binding.locationLayout.visibility = View.GONE
            }
            val timeCondition: TimeCondition? = profile.findCondition()
            if (timeCondition != null) {
                binding.days.text = timeCondition.daysActive.toDaysString(itemView.context)
                binding.interval.text = intervalToString(
                    timeCondition.startTime,
                    timeCondition.endTime
                )
                binding.timeLayout.visibility = View.VISIBLE
            } else {
                binding.timeLayout.visibility = View.GONE
            }
            val wifiCondition: WifiCondition? = profile.findCondition()
            if (wifiCondition != null) {
                binding.wifiSummary.text = wifiCondition.wifiList.map { it.ssid }.concatenate(", ")
                binding.wifiLayout.visibility = View.VISIBLE
            } else {
                binding.wifiLayout.visibility = View.GONE
            }
            val powerCondition: PowerCondition? = profile.findCondition()
            if (powerCondition != null) {
                binding.powerLayout.visibility = View.VISIBLE
                binding.powerSummary.setText(if (powerCondition.powerConnected) R.string.power_connected else R.string.power_disconnected)
            }
        }

        fun setSelected(selected: Boolean) {
            binding.cover.visibility = if (selected) View.VISIBLE else View.GONE
        }

        init {
            binding.root.setOnClickListener {
                listener.onProfileClick(adapterPosition)
            }
            binding.root.setOnLongClickListener {
                listener.onProfileLongClick(adapterPosition)
                true
            }
            this.listener = listener
        }
    }
}